### PR TITLE
Jmeter test automation implementation

### PR DESCRIPTION
**Pull request by sameerawickramasekara**
## Purpose
Resolves #4 

## Goals
 TestGrid will execute the tests in a specific deployment and solution pattern.

## User stories
Tests in the TESTGRID_HOME folder must be executed from the framework and generate a report from the result.

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Test environment
JDK 8, Ubuntu

### CODE REVIEW

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/beans/JmeterTest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import org.wso2.carbon.testgrid.automation.exceptions.TestGridExecuteException;
+import org.wso2.carbon.testgrid.automation.executers.common.TestExecuter;
+import org.wso2.carbon.testgrid.automation.executers.common.TestExecuterFactory;
+import org.wso2.carbon.testgrid.common.Deployment;
+import org.wso2.carbon.testgrid.common.constants.TestGridConstants;
+
+import java.util.List;
+
+/**
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/beans/JmeterTest.java]: Change the class comment. Also change the class name  and jmterScripts variable to have proper camel-notation.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/beans/TestNGTest.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+package org.wso2.carbon.testgrid.automation.beans;
+
+import org.wso2.carbon.testgrid.automation.exceptions.TestGridExecuteException;
+import org.wso2.carbon.testgrid.common.Deployment;
+
+/**
+ * This is the bean class for TestNG tests.
+ */
+public class TestNGTest extends Test {
+    @Override
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/beans/TestNGTest.java]: Leave a space between class declaration & methods.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/core/TestManager.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+     * @throws TestManagerException When there is an error creating the file structure.
+     */
+    public void init(String testLocation, Deployment deployment) throws TestManagerException {
+        this.deployment = deployment;
+        TestGridTestReader testGridTestReader = new TestGridTestReader();
+
+        try {
+            this.testLocation = testLocation + File.separator + "Tests";
+            this.tests = testGridTestReader.getTests(this.testLocation);
+
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/core/TestManager.java]: Remove this space.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/exceptions/TestGridExecuteException.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.testgrid.automation.exceptions;
+
+/**
+ * Exceptions occuring form the TestGrid execution.
+ */
+public class TestGridExecuteException extends Exception {
+    private static final long serialVersionUID = -3155676311329070297L;
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/exceptions/TestGridExecuteException.java]: Add a space here.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/executers/JmeterExecuter.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * This class is responsible for performing the tasks related to execution of single jmeter solution.
+ */
+public class JmeterExecuter implements TestExecuter {
+
+    private static final Log log = LogFactory.getLog(JmeterExecuter.class);
+    private String jmterHome;
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/executers/JmeterExecuter.java]: For readability change this variable to jMeterHome. I think we should get this from an environment variable.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/executers/common/TestExecuter.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+package org.wso2.carbon.testgrid.automation.executers.common;
+
+import org.wso2.carbon.testgrid.automation.exceptions.TestGridExecuteException;
+import org.wso2.carbon.testgrid.common.Deployment;
+
+/**
+ * Common interface for Test executer.
+ */
+public interface TestExecuter {
+    void execute(String script, Deployment deployment) throws TestGridExecuteException;
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/executers/common/TestExecuter.java]: Add a space here.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/file/TestNGTestReader.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import org.wso2.carbon.testgrid.automation.exceptions.TestReaderException;
+import org.wso2.carbon.testgrid.automation.file.common.TestReader;
+
+import java.util.List;
+
+/**
+ * This class is responsible for reading testNG tests for TestGrid framework.
+ */
+public class TestNGTestReader implements TestReader {
+    @Override
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/file/TestNGTestReader.java]: Add a space here.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/file/common/TestGridTestReader.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    /**
+     * This method goes thorugh every type of test in the folder structure and returns a list of tests
+     * with common Test interface.
+     *
+     * @param testLocation The file path of the test location as a String.
+     * @return a List of Tests.
+     * @throws TestReaderException when there is an error with test reading process.
+     */
+    public List<Test> getTests(String testLocation) throws TestReaderException {
+        File tests = new File(testLocation);
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/file/common/TestGridTestReader.java]: Remove unnecessary spaces inside this method.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/file/common/TestReaderFactory.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public class TestReaderFactory {
+    /**
+     * This method returns the TestReader implementation of the given type.
+     *
+     * @param testType Type of the tests as a String.
+     * @return object of TestReader.
+     */
+    public static TestReader getTestReader(String testType) {
+        switch (testType) {
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/file/common/TestReaderFactory.java]: Remove unnecessary spaces inside this method.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/file/common/TestReaderFactory.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.wso2.carbon.testgrid.automation.file.JmeterTestReader;
+import org.wso2.carbon.testgrid.automation.file.TestNGTestReader;
+import org.wso2.carbon.testgrid.common.constants.TestGridConstants;
+
+/**
+ * The factory class to get a Test reader implementation.
+ */
+public class TestReaderFactory {
+    /**
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/file/common/TestReaderFactory.java]: Add a space here.

File: utils/src/main/java/org/wso2/carbon/testgrid/utils/EnvVariableUtil.java [hunk 11]
```diff
@@ -0,0 +1,10 @@
+package org.wso2.carbon.testgrid.utils;
+
+/**
+ * Created by sameera on 30/10/17.
+ */
+public class EnvVariableUtil {
+    public static String readEnvironmentVariable(String variable){
```
Comment (harshanL) [inline utils/src/main/java/org/wso2/carbon/testgrid/utils/EnvVariableUtil.java]: Add a space here and update the class comment.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/TestEngine.java [hunk 12]
```diff
@@ -18,19 +18,74 @@
...
+import org.wso2.carbon.testgrid.common.Port;
 import org.wso2.carbon.testgrid.common.TestScenario;
+import org.wso2.carbon.testgrid.reporting.ReportingException;
+import org.wso2.carbon.testgrid.reporting.TestReportEngine;
 
-/**
- * Created by harshan on 10/30/17.
+import java.util.Arrays;
+
+/**This class is Responsible for initiating the core processes of TestGrid.
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/TestEngine.java]: Please fix the class comment format.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/TestEngine.java [hunk 13]
```diff
@@ -18,19 +18,74 @@
...
  */
 public class TestEngine {
 
-    private boolean getScenarioTests(String repo) {
-        return false;
-    }
+    private static final Log log = LogFactory.getLog(TestEngine.class);
 
+    /**
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/TestEngine.java]: Missing method comment.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/TestEngine.java [hunk 14]
```diff
@@ -18,19 +18,74 @@
...
+            scenario.setDeployment(deployment);
+            testManager.init(scenario.getScenarioLocation(), deployment);
+            testManager.executeTests();
+
+            TestReportEngine engine = new TestReportEngine();
+            engine.generateReport(scenario);
+
+        } catch (TestManagerException ex) {
+            String msg = "Error while initiating the TestManager";
+            log.error(msg, ex);
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/TestEngine.java]: Do not log and throw an exception unless it's an exit point of an API.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/beans/JMeterTest.java [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+    private List<String> getjMeterScripts() {
+        return jMeterScripts;
+    }
+
+    public void setjMeterScripts(List<String> jMeterScripts) {
+        this.jMeterScripts = jMeterScripts;
+    }
+
+    /**
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/beans/JMeterTest.java]: Missing method comment.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/core/TestManager.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+ * TestManager class manages test executions of all the test types in the test grid folder.
+ */
+public class TestManager {
+
+    private static final Log log = LogFactory.getLog(TestManager.class);
+    private List<Test> tests;
+    private Deployment deployment;
+    private String testLocation;
+
+    /**
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/core/TestManager.java]: Missing method comment.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/executers/JMeterExecuter.java [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+     */
+    @Override
+    public void execute(String script, Deployment deployment) throws TestGridExecuteException {
+        String scriptName = getScriptName(script);
+        Runtime runtime = Runtime.getRuntime();
+        try {
+            String prop = changePropertyFile(deployment);
+            String command;
+            if (prop != null) {
+                command = "sh " + this.jMeterHome + "/bin/jmeter.sh -n -t " + script + " -l " +
```
Comment (harshanL) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/executers/JMeterExecuter.java]: We can use a String Builder to construct the command string.

File: automation/src/main/java/org/wso2/carbon/testgrid/automation/executers/JMeterExecuter.java [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+    /**
+     * This method is responsible for updating the jmeter properties file with current deployment.
+     *
+     * @param deployment The deployment details of the current pattern.
+     * @return the path of the modified user.properties file of jmeter test.
+     * @throws ConfigurationException when there is an error reading the user.properties file.
+     */
+    private String changePropertyFile(Deployment deployment) throws ConfigurationException {
+        File file = new File(testGridFolder + File.separator + "JMeter" + File.separator + testName +
+                "/src/test/resources/user.properties");
```
Comment (lasanthaDLPDS) [inline automation/src/main/java/org/wso2/carbon/testgrid/automation/executers/JMeterExecuter.java]: Shall we use Java Path interface to get the file path

File: utils/src/main/java/org/wso2/carbon/testgrid/utils/GitUtil.java [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+
+    /**
+     *
+     * @param fileLocation destination directory of tests
+     * @param gitLocation git URL of test repo
+     * @return the File object with tests
+     * @throws GitAPIException when there is an error in cloning process
+     */
+    public static File cloneGitRepo(String fileLocation,String gitLocation) throws GitAPIException {
+
```
Comment (vidurananayakkara) [inline utils/src/main/java/org/wso2/carbon/testgrid/utils/GitUtil.java]: An unrequired blank line here. Shall we remove it?

File: utils/src/main/java/org/wso2/carbon/testgrid/utils/GitUtil.java [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+        File testDirectory = new File(fileLocation);
+        Git.cloneRepository().setURI(gitLocation).setDirectory(testDirectory).call();
+        return testDirectory;
+
```
Comment (vidurananayakkara) [inline utils/src/main/java/org/wso2/carbon/testgrid/utils/GitUtil.java]: An unrequired blank line here. Shall we remove it?

**Pull request merged**